corehq/apps/users/forms.py

```diff
@@ -768,13 +777,70 @@ def __init__(self, *args, **kwargs):
         self.helper.label_class = 'col-sm-3 col-md-2'
         self.helper.field_class = 'col-sm-9 col-md-8 col-lg-6'
 
+    @property
+    @memoized
+    def commtrack_enabled(self):
+        return Domain.get_by_name(self.domain).commtrack_enabled
+
     def save(self, user):
+        # todo: Avoid multiple user.save
+        domain_membership = user.get_domain_membership(self.domain)
+        if self.commtrack_enabled:
+            domain_membership.program_id = self.cleaned_data['program_id']
+
+        self._update_location_data(user)
+
+    def _update_location_data(self, user):
         location_id = self.cleaned_data['primary_location']
-        if location_id:
-            if location_id != user.location_id:
-                user.set_location(Location.get(location_id))
+        location_ids = self.cleaned_data['assigned_locations']
+
+        if user.is_commcare_user():
+            old_location_id = user.location_id
+            if location_id != old_location_id:
+                if location_id:
+                    user.set_location(Location.get(location_id))
+                else:
+                    user.unset_location()
+
+            old_location_ids = user.assigned_location_ids
+            if set(location_ids) != set(old_location_ids):
+                if location_ids:
+                    user.assign_to_locations(location_ids)
+                else:
+                    user.unassign_all_locations()
         else:
-            user.unset_location()
+            domain_membership = user.get_domain_membership(self.domain)
+            old_location_id = domain_membership.location_id
+            if location_id != old_location_id:
+                if location_id:
+                    user.set_location(self.domain, Location.get(location_id))
+                else:
+                    user.unset_location(self.domain)
+
+            old_location_ids = domain_membership.assigned_location_ids
+            if set(location_ids) != set(old_location_ids):
+                if location_ids:
+                    user.assign_to_locations(self.domain, location_ids)
+                else:
+                    user.unassign_all_locations(self.domain)
+
+    def clean_assigned_locations(self):
+        # select2 (< 4.0) doesn't format multiselect for remote data as an array
+        #   but formats it as comma-seperated list, so we need to clean the returned data
+        from corehq.apps.locations.models import SQLLocation
+        from corehq.apps.locations.util import get_locations_from_ids
+
+        value = self.cleaned_data.get('assigned_locations')
+        if not isinstance(value, basestring) or value.strip() == '':
+            return []
+
+        location_ids = [location_id.strip() for location_id in value.split(',')]
```

Comment:
Any chance select2 could give empty values or use a trailing comma here?  If so, it might be prudent to wrap this in a `filter(None, ...)` to drop null values.